extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/StaticTenantResolver.java

```diff
@@ -0,0 +1,287 @@
+                TenantConfigContext defaultTenant) {
+            final var builder = ImmutablePathMatcher.<String> builder().rootPath(rootPath);
+            addPath(DEFAULT_TENANT, defaultTenant.oidcConfig(), builder);
+            for (Map.Entry<String, TenantConfigContext> e : staticTenantsConfig.entrySet()) {
+                addPath(e.getKey(), e.getValue().oidcConfig(), builder);
+            }
+            return builder.hasPaths() ? new PathMatchingTenantResolver(builder.build()) : null;
+        }
+
+        @Override
+        public String resolve(RoutingContext context) {
+            String tenantId = staticTenantPaths.match(context.normalizedPath()).getValue();
+            if (tenantId != null) {
+                LOG.debugf(
+                        "Tenant id '%s' is selected on the '%s' request path", tenantId, context.normalizedPath());
+                return tenantId;
+            }
+            return null;
+        }
+
+        private static ImmutablePathMatcher.ImmutablePathMatcherBuilder<String> addPath(String tenant, OidcTenantConfig config,
+                ImmutablePathMatcher.ImmutablePathMatcherBuilder<String> builder) {
+            if (config != null && config.tenantPaths.isPresent()) {
+                for (String path : config.tenantPaths.get()) {
+                    builder.addPath(path, tenant);
+                }
+            }
+            return builder;
+        }
+    }
+
+    private static final class IssuerBasedTenantResolver {
+
+        private final TenantConfigContext[] tenantConfigContexts;
+        private final boolean detectedTenantWithoutMetadata;
+        private final Map<String, AtomicBoolean> tenantToRetry;
+
+        private IssuerBasedTenantResolver(TenantConfigContext[] tenantConfigContexts, boolean detectedTenantWithoutMetadata,
+                Map<String, AtomicBoolean> tenantToRetry) {
+            this.tenantConfigContexts = tenantConfigContexts;
+            this.detectedTenantWithoutMetadata = detectedTenantWithoutMetadata;
+            this.tenantToRetry = tenantToRetry;
+        }
+
+        private Uni<String> resolveTenant(RoutingContext context) {
+            return resolveTenant(context, 0);
+        }
+
+        private Uni<String> resolveTenant(RoutingContext context, int index) {
+            if (index == tenantConfigContexts.length) {
+                return Uni.createFrom().nullItem();
+            }
+            var tenantContext = tenantConfigContexts[index];
+            if (detectedTenantWithoutMetadata) {
+                // this is static tenant that didn't have OIDC metadata available at startup
+
+                if (tenantContext.getOidcMetadata() == null) {
+                    if (tenantContext.ready()) {
+                        return resolveTenant(context, index + 1);
+                    }
+
+                    if (!tryToInitialize(tenantContext)) {
```

Comment:
@sberyozkin these are the drons you are looking for

Reply:
Thanks @michalvavrik, I see it now.

I believe this map though should be at the `StaticTenantResolver` level, for `DynamicTenantConfigResolver#initializeStaticTenantContextIfNotReady` to be able to use it as well.
Otherwise, non-ready static tenants resolved from other static tenant resolvers will keep stressing the system

Reply:
The thing is, that static tenants could keep pressing system because they were simply selected by some tenant resolver. In which case they are desired.

Reply:
sorry if I am being to careful

Reply:
Hi Michal, @michalvavrik,  one thing we definitely agree upon is that fixing it for the issuer based resolver is necessary, and indeed, let's not try to put the unrelated change into this PR by extending this restriction to all other static tenants.
I'll open another issue to handle it in another PR.

>until now, it was possible for all the static tenants to recover anytime (after 20 initialization failures, even 1000).

Yeah, but it was never intended to be more than a single init retry, I only realized it was the case while discussing the issuer based resolver with yourself here. Right now it is even more than 1000 :-), it is an indefinite number of retries across possibly a highly concurrent system. The OIDC provider might not be available at the startup, but if after another retry at the startup, it is still not available, then I think just keeping returning 401 is cheaper than failing to connect forever and returning 401.

But indeed, let's finalize this PR, thanks

Reply:
> The OIDC provider might not be available at the startup, but if after another retry at the runtime, it is still not available, then I think just keeping returning 401 is cheaper than failing to connect forever and returning 401.

I agree with that and I am happy to participate on changes whether in this or about finish encapsulation of dynamic tenants or whatever is needed. I was just worried to do it here. Thanks for insights on intended behavior.